instrumentation/net_http/lib/opentelemetry/instrumentation/net/http/patches/instrumentation.rb

```diff
@@ -70,6 +69,14 @@ def annotate_span_with_response!(span, response)
             def tracer
               Net::HTTP::Instrumentation.instance.tracer
             end
+
+            def config
+              Net::HTTP::Instrumentation.instance.config
+            end
+
+            def uri_from_req(req)
+              req.uri || URI.join("#{USE_SSL_TO_SCHEME[use_ssl?]}://#{@address}:#{@port}", req.path)
```

Comment:
This is awkward - we're piecing together the URL string so the helper can pull it apart, which is somewhat perversely inefficient when we already have the fragments we need in hand. IIRC we did a bunch of work to reduce allocations & string manipulation as much as possible in this instrumentation.

Reply:
I think this is from one of my commits on the earlier PR. I'm a bit rusty on this since it's been a while, but I think my reasoning can be summarised as:
- req.uri is not guaranteed to be present as it also accepts a path only parameter which is a string
- I wanted to set the `http.url` so that it's consistent across clients. The services teams run have a few different http clients as dependencies and I find it really annoying not knowing which attributes to use when searching for a particular endpoint's client spans.
- The original PR added a hide query params option to emit query params as they often include sensitive information.  This was done using the `URI` class' `#query=` method.

I think that I probably would be happy about not setting `http.url` if we are consistently setting the `http.target` using `uri.path, `uri.request_uri`, or `path` as that should always be possible.

If we choose to not construct a URI then we can probably do something by splitting the path String at `&` to omit query params

Reply:
i tried to clean the request attributes helper up to both merge in the http client context attribs automagically, and also expose some more optional arguments so that it's flexible enough to not have to construct a URI in the case where it doesn't already exist.